internal/generator/vector/output/loki/loki.go

```diff
@@ -0,0 +1,221 @@
+package loki
+
+import (
+	"fmt"
+	"strings"
+
+	logging "github.com/openshift/cluster-logging-operator/apis/logging/v1"
+	"github.com/openshift/cluster-logging-operator/internal/constants"
+	. "github.com/openshift/cluster-logging-operator/internal/generator"
+	. "github.com/openshift/cluster-logging-operator/internal/generator/vector/elements"
+	vectorhelpers "github.com/openshift/cluster-logging-operator/internal/generator/vector/helpers"
+	"github.com/openshift/cluster-logging-operator/internal/generator/vector/output/security"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const (
+	lokiLabelKubernetesHost = "kubernetes.host"
+)
+
+var (
+	defaultLabelKeys = []string{
+		"log_type",
+
+		//container labels
+		"kubernetes.namespace_name",
+		"kubernetes.pod_name",
+		"kubernetes.container_name",
+	}
+	requiredLabelKeys = []string{
+		lokiLabelKubernetesHost,
+	}
+	lokiEncodingJson = fmt.Sprintf("%q", "json")
+)
+
+type Loki struct {
+	ComponentID string
+	Inputs      string
+	TenantID    Element
+	Endpoint    string
+	LokiLabel   []string
+}
+
+func (l Loki) Name() string {
+	return "lokiVectorTemplate"
+}
+
+func (l Loki) Template() string {
+	return `{{define "` + l.Name() + `" -}}
+[sinks.{{.ComponentID}}]
+type = "loki"
+inputs = {{.Inputs}}
+endpoint = "{{.Endpoint}}"
+{{kv .TenantID -}}
+{{end}}`
+}
+
+type LokiEncoding struct {
+	ComponentID string
+	Codec       string
+}
+
+func (le LokiEncoding) Name() string {
+	return "lokiEncoding"
+}
+
+func (le LokiEncoding) Template() string {
+	return `{{define "` + le.Name() + `" -}}
+[sinks.{{.ComponentID}}.encoding]
+codec = {{.Codec}}
+{{end}}`
+}
+
+type Label struct {
+	Name  string
+	Value string
+}
+
+type LokiLabels struct {
+	ComponentID string
+	Labels      []Label
+}
+
+func (l LokiLabels) Name() string {
+	return "lokiLabels"
+}
+
+func (l LokiLabels) Template() string {
+	return `{{define "` + l.Name() + `" -}}
+[sinks.{{.ComponentID}}.labels]
+{{range $i, $label := .Labels -}}
+{{$label.Name}} = "{{$label.Value}}"
+{{end -}}
+{{end}}
+`
+}
+
+func Conf(o logging.OutputSpec, inputs []string, secret *corev1.Secret, op Options) []Element {
+	return MergeElements(
+		[]Element{
+			Output(o, inputs, secret, op),
+			Encoding(o),
+			Labels(o),
+		},
+		TLSConf(o, secret),
+		BasicAuth(o, secret),
+	)
+}
+
+func Output(o logging.OutputSpec, inputs []string, secret *corev1.Secret, op Options) Element {
+	return Loki{
+		ComponentID: strings.ToLower(vectorhelpers.Replacer.Replace(o.Name)),
+		Inputs:      vectorhelpers.MakeInputs(inputs...),
+		Endpoint:    o.URL,
+		TenantID:    Tenant(o.Loki),
+	}
+}
+
+func Encoding(o logging.OutputSpec) Element {
+	return LokiEncoding{
+		ComponentID: strings.ToLower(vectorhelpers.Replacer.Replace(o.Name)),
+		Codec:       lokiEncodingJson,
+	}
+}
+
+func lokiLabelKeys(l *logging.Loki) []string {
+	var keys sets.String
+	if l != nil && len(l.LabelKeys) != 0 {
```

Comment:
can l.LabelKeys be `null` and len method may fail ?

Reply:
length of a `nil` slice is 0
https://go.dev/play/p/knVjnFRAiBb